subprojects/jacoco/src/main/java/org/gradle/internal/jacoco/transform/AgentJarExtraction.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.jacoco.transform;
+
+import org.gradle.api.GradleException;
+import org.gradle.api.UncheckedIOException;
+import org.gradle.api.artifacts.transform.InputArtifact;
+import org.gradle.api.artifacts.transform.TransformAction;
+import org.gradle.api.artifacts.transform.TransformOutputs;
+import org.gradle.api.artifacts.transform.TransformParameters;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.provider.Provider;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.Enumeration;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+public abstract class AgentJarExtraction implements TransformAction<TransformParameters.None> {
+
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInputArtifact();
+
+    @Override
+    public void transform(TransformOutputs outputs) {
+        File input = getInputArtifact().get().getAsFile();
+        File output = outputs.file(input.getName());
```

Comment:
I'm reusing the input file name so that the version matching code in `JacocoAgentJar` continues to work. I think the version meta data should be transfered in some other way, but I don't know how. Should this transform write a text file containing the version?

Reply:
I think this is a different topic. Looking at the code the version was always extracted from the file name. So if we keep that, I think what you are doing is fine.

Another (better?) approach would be to pass the version, that was configured on the extension, to the `AgentJar` class when it is created and then use that. Because the version is already used to define the dependency on the `agentConf` which then triggers the transform. So if we know the version then, why do we need to extract it again later?